### DISCUSSION TITLE
Type parameter when returning Scope with middleware from fn

### DISCUSSION THREAD

**Discussion by Tipnos** [actix/actix-web#3074] [Q&A] [answered]
## Context

Route configuration of a service `Scope` changes regards to a `clap` argument when starting an actix server.

The route configuration is implemented in a function which takes the `clap` argument as param and return the corresponding `Scope`.  Everything works fine as long as no middleware is wrapped inside the function.

The simple example below trigger the error.

## Example

```rust 
use actix_web::{http::Method, middleware, web, App, HttpResponse, HttpServer};

#[actix_web::main]
async fn main() -> std::io::Result<()> {
    HttpServer::new(|| {
        App::new()
            .service(web::resource("/health-check").route(web::get().to(HttpResponse::Ok)))
            .service(api_service(12))
    })
    .bind(("127.0.0.1", 8080))?
    .run()
    .await
}

fn api_service(arg_from_clap: i64) -> actix_web::Scope {
    actix_web::Scope::new("/api")
        .wrap(middleware::DefaultHeaders::new().add(("X-Version", "0.2"))) // Example compile if this line is removed
        .service(
            web::resource("/test")
                .route(web::get().to(HttpResponse::Ok))
                .route(web::method(Method::HEAD).to(HttpResponse::MethodNotAllowed)),
        )
} 
```

## Compiler error

`expected struct 'actix_web::Scope<actix_web::scope::ScopeEndpoint>'
               found struct 'actix_web::Scope<impl ServiceFactory<ServiceRequest, Config = (), Response = ServiceResponse, Error = actix_web::Error, InitError = ()>>'`
               
## Version
actix-web = 4
rustc = 1.69.0

**Comment by robjtede** [comment, marked as answer]
To change the default type parameter for `Scope`s:

```
fn api_service(arg_from_clap: i64) -> actix_web::Scope<
    impl ServiceFactory<
        ServiceRequest,
        Response = ServiceResponse<impl MessageBody>,
        Config = (),
        InitError = (),
        Error = actix_web::Error,
    >,
> {
    // ...
}
```
